File: Ex8/readtria.py
#!/usr/bin/python3
# Note: Code probably needs to be changed if extra marker are introduced
import numpy as np
import matplotlib.pyplot as plt

def readtria(fname):
    # input: filename
    # output: x,y,npoint,nelement,e2p,idp,ide
    print("read mesh file: " + fname)  
    
    # Node file
    npoint = ndim = npattr = nbound = None    
    fname_node = fname + ".node"
    with open(fname_node, "r") as fnode:    
        npoint, ndim, npattr, nbound = [int(x) for x in fnode.readline().split()]         

    U = np.genfromtxt(fname_node, skip_header=1)
    x = U[:,1]               # dimension
    y = U[:,2]               # number of point attributes
    idp = U[:,3].astype(int) # number of boundary markers

    
    
    # Element file
    nelement = nphi = neattr = None    
    fname_ele = fname + ".ele"
    with open(fname_ele, "r") as fele:
        nelement, nphi, neattr = [int(x) for x in fele.readline().split()]         

    U = np.genfromtxt(fname_ele, skip_header=1, dtype=np.int64)
    e2p = U[:, 1:nphi+1] -1
    ide = U[:, nphi+1:]
    
    return x,y,npoint,nelement,e2p,idp,ide


x,y,npoint,nelement,e2p,idp,ide = readtria("box.2")
plt.scatter(x,y)
for tri in e2p:
    plt.plot([x[tri[0]], x[tri[1]]], [y[tri[0]], y[tri[1]]])
    plt.plot([x[tri[1]], x[tri[2]]], [y[tri[1]], y[tri[2]]])
    plt.plot([x[tri[0]], x[tri[2]]], [y[tri[0]], y[tri[2]]])
plt.show()
print("Number of triangular elements: ", int(len(e2p)/3))
print("Number of nodes: ", len(x))

File: Ex8/readtria.m
% read a mesh from triangle
% sample code by Dirk Peschka

% Note: Code probably needs to be changed if extra marker are introduced

function [x,y,npoint,nelement,e2p,idp,ide] = readtria(fname)

fnode = strcat(fname,'.node');
fele  = strcat(fname,'.ele');

fprintf('read mesh file: %s \n',fname);

% read .node file
fid = fopen(fnode,'r');
mesh_header = fscanf(fid,'%i',4);

npoint = mesh_header(1); % number of points
ndim   = mesh_header(2); % dimension
npattr = mesh_header(3); % number of point attributes
nbound = mesh_header(4); % number of boundary markers

x   = zeros(npoint,1);
y   = zeros(npoint,1);
idp = zeros(npoint,1);

for i=1:npoint
    dump  =fscanf(fid,'%i',1); % index running from 1..npoint
    x(i)  =fscanf(fid,'%f',1); % x-position
    y(i)  =fscanf(fid,'%f',1); % y-position
    idp(i)=fscanf(fid,'%i',1); % boundary marker
end
fclose(fid);

% read .ele file
fid = fopen(fele,'r');
mesh_header = fscanf(fid,'%i',3);

nelement = mesh_header(1); % number of points
nphi     = mesh_header(2); % dimension
neattr   = mesh_header(3); % number of element attributes

e2p = zeros(nelement,nphi);
ide = zeros(nelement,neattr);

for k=1:nelement
    dump    =fscanf(fid,'%i',1); % index running from 1..nelement
    for j=1:nphi
        e2p(k,j)=fscanf(fid,'%i',1); % element-to-point 1..nphi
    end
    for j=1:neattr
        ide(k,j)=fscanf(fid,'%i',1); % element-marker 1..neattr
    end
end
fclose(fid);

npoint = max(max(e2p(:,1:3))); % use npoint for P1 points